src/constraint_system/logic.rs

```diff
@@ -248,13 +248,22 @@ impl StandardComposer {
         // x & ((1 << num_bits +1) -1) == [0..num_quads] accumulated sums of x.
         //
         // We could also check that the last gates wire coefficients match the
-        // original values introduced in the function.
+        // original values introduced in the function taking into account the bitnum
+        // specified on the fn call parameters.
         // This can be done with an `assert_equal` constraint gate or simply
         // by taking the values behind the n'th variables of `w_l` & `w_r` and
         // checking that they're equal to the original ones behind the variables
         // sent through the function parameters.
-        assert_eq!(self.variables[&a], self.variables[&self.w_l[self.n - 1]]);
-        assert_eq!(self.variables[&b], self.variables[&self.w_r[self.n - 1]]);
+        assert_eq!(
+            self.variables[&a]
+                & (Scalar::from(2u64).pow(&[(num_bits) as u64, 0, 0, 0]) - Scalar::one()),
```

Comment:
Since I start to see few places where we use this, what about having:

```rust
self.variables[&a] & Scalar::pow_of_2(num_bits) - Scalar::one(),
```

instead?
Not only it's more concise and readable, but the implementation can remove the unnecessary iteration and the array of four `u64` creation in the stack since we just pass a single value (`usize` I assume).

Reply:
Hmmm. I din't want to introduce much more new methods like this one. At the end, you can already do the same thing just with the `pow` method.

IMO it's not worth it.

Reply:
That would also applying to all the helper methods we already introduced: you could do the same thing (jubjub to bls, checking if is zero, and so on) so I don't see why not at this point. Our code already diverged substantially from the original repo (and not just because the helper method) so it's not like this change would impact that.

As explained the difference between using `pow` and a dedicated method is also not just in readability but in unnecessary operation involved: you pass an array (unnecessary) that you're going to fully iterate 4 times (unnecessary) every time you just want a scalar that is a power of two.

Other libraries have similar methods – if not the same – also because in general can be optimized differently from a `pow` (but even without specific optimization, just the same code of `pow` will be reduced).

Reply:
Let's open a low prio issue for it if you want. But i don't think it has to be solved in this PR.

Reply:
Done, See https://github.com/dusk-network/plonk/issues/259